src/nn/transformer.rs

```diff
@@ -0,0 +1,1097 @@
+                    -0.0994, -0.1629, -0.2694, -0.0869, 0.1631, -0.1892, -0.0901, 0.3148,
+                ]),
+            },
+            w_o: Linear {
+                weight: Tensor2D::new([
+                    [
+                        0.1190, -0.2099, 0.1869, -0.3508, 0.0826, -0.3263, 0.2366, -0.2100,
+                    ],
+                    [
+                        0.2002, -0.2365, -0.1015, 0.2539, -0.1125, 0.2926, -0.0981, -0.1495,
+                    ],
+                    [
+                        -0.1831, 0.0348, 0.2623, 0.1650, 0.2114, -0.0376, 0.1850, -0.3326,
+                    ],
+                    [
+                        -0.0636, 0.1737, -0.1024, -0.0246, 0.2178, -0.3127, -0.0506, 0.0568,
+                    ],
+                    [
+                        -0.3384, -0.1202, -0.2316, 0.0117, 0.2929, -0.2060, 0.1966, -0.3274,
+                    ],
+                    [
+                        0.2589, 0.3003, -0.2277, 0.2488, -0.0594, -0.0645, 0.0931, 0.2376,
+                    ],
+                    [
+                        0.3371, 0.0463, -0.1292, 0.1341, 0.2008, -0.0325, 0.0914, 0.0517,
+                    ],
+                    [
+                        -0.2241, 0.0426, 0.2326, -0.3048, -0.2760, -0.0868, -0.2429, 0.1446,
+                    ],
+                ]),
+                bias: Tensor1D::new([
+                    0.0800, 0.0567, 0.2609, -0.1651, -0.0820, -0.1058, -0.3133, -0.1181,
+                ]),
+            },
+        };
+        let x: Tensor2D<2, 8> = Tensor2D::new([
+            [
+                0.7207, 0.3572, 0.2341, 0.4865, 0.2949, 0.5450, 0.8236, 0.4674,
+            ],
+            [
+                0.4800, 0.6774, 0.9052, 0.4714, 0.5683, 0.7339, 0.1975, 0.3909,
+            ],
+        ]); // Sequence of 2 token vectors with 8 dims each
+        let y: Tensor2D<2, 8> = model.forward(x);
+        assert_close(
+            y.data(),
+            &[
+                [
+                    0.3727123,
+                    -0.06548928,
+                    0.1931893,
+                    0.015571773,
+                    0.16583481,
+                    -0.073905066,
+                    -0.19729866,
+                    -0.21708608,
+                ],
+                [
+                    0.37352866,
+                    -0.067618236,
+                    0.19414166,
+                    0.016556486,
+                    0.16638368,
+                    -0.074163824,
+                    -0.19822657,
+                    -0.21566412,
+                ],
+            ],
+        );
+    }
+
+    #[test]
+    fn test_matmul() {
+        let mut rng = thread_rng();
+        let inp: Tensor1D<4> = Tensor1D::rand(&mut rng);
+        let out: Tensor2D<2, 2> = inp.duplicate().reshape();
+        println!("Inp: {:?}", inp);
+        println!("Out: {:?}", out);
+        //panic!("")
+    }
+
+    #[test]
+    fn test_transformer_encoder() {
```

Comment:
I've been meaning to do this for a while, but i'd like to take all the actual floats out of the tests so they are easier to scan. Don't need to address in this PR, but probably shortly after. I think we can take advantage of the numpy::load functions here and store .npy files somewhere in the repo.

Reply:
I like that idea. I'm going to leave it for a seperate PR though.